### PR TITLE
build: Add make nconfig toplevel target

### PR DESCRIPTION
**Pull request by alive4ever**
Add ``make nconfig`` toplevel target to make use of nconf UI as found on kernel's build system.
lede-nconfig

### CODE REVIEW

**Comment by alive4ever** [comment]
Rebased to latest HEAD

**Comment by alive4ever** [comment]
Rebased to latest HEAD.

I really want this to get merged, so `nconfig` ui can be used in addition to the good old `menuconfig`.

Reasons to use `make nconfig` over `make menuconfig`

* Better navigation
Arrow keys are better utilized for navigation. Instead of pressing `Enter` to follow the submenu, `right arrow` can be used. Similarly, instead of choosing `Exit` at every submenu to navigate back to upper menu, `left key` can be used.
`pg-up` and `pg-down` keys can be used to scroll the display on all menus. On `menuconfig`, the `pg-up` and `pg-down` keys cannot be used to scroll the display on `Target System` submenu.

* Better search
Incremental search using `/` key to quickly select the desired menu. The incremental search feature works on all menu display. In comparation, `menuconfig` search doesn't do incremental search selection. It only does symbol search.

**Comment by alive4ever** [comment]
Ping @pepe2k , @diizzyy , @nbd168 , and other reviewers.

**Comment by diizzyy** [comment]
Looks pretty nice! I'm a bit worried though however how well maintained this is upstream. 

As for the search features, does that mean that you can search for packages (and options) directly anywhere?

**Comment by alive4ever** [comment]
> As for the search features, does that mean that you can search for packages (and options) directly anywhere?

Yeah. Try it out
```
git remote add alive4ever https://github.com/alive4ever/source
git fetch alive4ever
git checkout make-nconfig
make nconfig
```

**Comment by zorun** [comment]
This is nice, I used nconf not long ago in another project and it's indeed an improvement over menuconfig.

However your branch does not build on Debian stretch x86_64, see: https://paste.aliens-lyon.fr/X1y
It seems that you are missing some linking flags against ncurses, and there are some strange errors on top.

**Comment by alive4ever** [comment]
On those Debian based distributions, you need to install distributions's `pkg-config` to use `make nconfig`.

This is required to use find and use the host-provided tools and development libraries, since `ncurses` is not part of `hostpkg` on LEDE build system.

**Comment by zorun** [comment]
Ok thanks!  I'm wondering why pkg-config is needed for this, because ncurses was already required before (not sure exactly why), but not pkg-config.

By the way, are you sure the new check is working?  Because even without pkg-config, the check seems to succeed:

```
Checking 'working-make'... ok.
Checking 'case-sensitive-fs'... ok.
...
Checking 'git'... ok.
Checking 'file'... ok.
Checking 'pkg-config'... ok.
Checking 'ldconfig-stub'... ok.
```

Also, the nconfig target fails, but without the error message you added:
```
$ make nconfig
make: scripts/config/nconf: Command not found
/home/build/lede/include/toplevel.mk:145: recipe for target 'nconfig' failed
make: *** [nconfig] Error 127
```

It is only visible with `V=s`, while I think it should be shown without:
```
$ make nconfig V=s
make[1]: Entering directory '/home/build/lede/scripts/config'
You don't have pkg-config installed. Please install pkg-config
For now, you can use 'make menuconfig' instead of 'make nconfig'
make[1]: Leaving directory '/home/build/lede/scripts/config'
if [ \! -e .config -a -e /home/build/.openwrt/defconfig ]; then \
	cp /home/build/.openwrt/defconfig .config; \
fi
scripts/config/nconf Config.in
make: scripts/config/nconf: Command not found
/home/build/lede/include/toplevel.mk:145: recipe for target 'nconfig' failed
make: *** [nconfig] Error 127
```

**Comment by zorun** [comment]
Also, you should add the generate files to `scripts/config/.gitignore`, because git sees them:

```
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	scripts/config/nconf
	scripts/config/nconf_check
```

Other than that, after installing pkg-config and manually cleaning up `scripts/config/*.o`, it works fine, thanks :)

**Comment by zorun** [comment]
`
`
In fact, I have another issue: when switching back to master and trying to use the regular `menuconfig`, it fails:

```
$ make menuconfig V=s
make[1]: Entering directory '/home/build/lede/scripts/config'
/usr/bin/ld: lxdialog/checklist.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: lxdialog/util.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: lxdialog/inputbox.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: lxdialog/textbox.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: lxdialog/yesno.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: lxdialog/menubox.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
Makefile:32: recipe for target 'mconf' failed
```

Maybe you don't use the same build options?

**Comment by alive4ever** [comment]
> Other than that, after installing pkg-config and manually cleaning up scripts/config/*.o, it works fine, thanks :)

You should run `make -C ./scripts/config clean` to properly clean up the nconfig leftover objects and binary.

> In fact, I have another issue: when switching back to master and trying to use the regular menuconfig, it fails:

Do you have unusual system configuration, i.e. multiarch or similar?
This works fine for me with pure 64 bit Debian 9.1 container.

Consider cleaning up `./scripts/config` directory using `make -C ./scripts/config clean`.

**Comment by alive4ever** [comment]
For your information, the check of `pkg-config` requirement will place a symbolic link to `/usr/bin/pkg-config` inside `./staging_dir/host/bin` directory. You may remove the symlink if you are going back to the master branch after checking out my branch.

**Comment by alive4ever** [comment]
I shouldn't add pkg-config check during prereq, since the symbolic link of distro's pkg-config will conflict with `pkg-config` built during the build process.

I'll undo the relevant commit to fix the issue.

**Comment by diizzyy** [comment]
Hmm... there are a few interesting quirks

*  Keyboard mapping is a bit weird (using putty)
* F1-F4 can do anything from showing you generic help to exit nconfig (I once got F2 to display package info)
* F5+ seems to work as expected, F8 should probably be mapped to '/'
* Search '/' sometimes work? 
Color scheme needs to be adjusted ;-)

**Comment by nbd168** [comment]
Please rebase on top of the xconfig changes that I merged into my staging tree.

File: include/toplevel.mk [hunk 1]
```diff
@@ -136,6 +141,12 @@ menuconfig: scripts/config/mconf prepare-tmpinfo FORCE
 	[ -L .config ] && export KCONFIG_OVERWRITECONFIG=1; \
 		$< Config.in
 
+nconfig: scripts/config/nconf prepare-tmpinfo FORCE
+	if [ \! -e .config -a -e $(HOME)/.openwrt/defconfig ]; then \
```
Comment (pprindeville) [inline include/toplevel.mk]: Why do you need the `\!` instead of just `!`?

File: scripts/config/nconf.c [hunk 2]
```diff
@@ -0,0 +1,1557 @@
+/*
```
Comment (pprindeville) [inline scripts/config/nconf.c]: Can we pull `nconf.[hc]` from a URL instead of embedding them directly?

**Comment by blogic** [comment]
With the remerge in progress, all PRs on the lede-project organisation will be closed. Please help getting this merged or rebase/post it on the openwrt project page (https://github.com/openwrt/openwrt/pulls). All remaining PRs will be closed in 30 days.

**Comment by alive4ever** [comment]
Rebased with the latest codebase. Instead of using buildroot's nconf, use nconf sources from kernel 4.9.75 package.

File: scripts/config/nconf.c (hunk 2 above)
Comment (alive4ever) [reply scripts/config/nconf.c]: LEDE/OpenWrt build phases are configuration, source downloads, tools compilation, toolchain compilation, target compilation, and firmware generation.

Adding download during configuration phase doesn't make sense.

**Pull request closed without merging**